docs/specs/circuit.md

```diff
@@ -53,228 +53,219 @@ The chips that get to execute instructions are _instruction executors_. These ch
-receive) that must be equal at the end. To match the literature, let send (resp. receive) correspond to Write (resp.
-Read) sets. Any memory access in a chip must add one entry into each set and constrain a relation between them:
-
-- A read of $(a, v)$ at time $t$ must add $(a, v, t_{prev})$ to Read set and $(a, v, t)$ to Write set. It must constrain
-  $t_{prev} < t$.
-- A write of $(a, v)$ at time $t$ must add $(a,v_{prev},t_{prev})$ to Read set and $(a,v,t)$ to Write set, where $v_
-  {prev}$ is the previous value before the write. It must constrain $t_{prev} < t$.
-
-To balance the Read and Write sets, an additional chip must ensure that every accessed address has an initial $(a, v_
-{init}, 0)$ added to the Write set, and $(a, v_{final}, t_{last})$ added to Read set.
-
-<!--
-For the offline checking to be sound, it must be constrained that the list of accessed addresses in the initial Write
-list are all **unique**. Uniqueness of the initial address list implies, together with the bus argument, that the final
-address list is also unique (vice versa, uniqueness of final set implies uniqueness of initial set). The key observation
-of [OLB24](https://eprint.iacr.org/2024/979.pdf) is that only _uniqueness_ is necessary and not sorted-ness of the
-address list. The traditional approach prior to OLB24 to enforce uniqueness is to enforce the list of addresses is
-sorted, which uses logup lookups for range checks necessary to constrain `IsLessThan`. OLB24 shows that one can
-implement an AIR with in-circuit randomness to constrain that all entries in a trace column are unique (with an
-extension to conditional uniqueness).
--->
+## Adapter-core interface
+
+The adapter-core interface is defined by the `VmAdapterInterface` trait.
+
+```rust
+/// The interface between primitive AIR and machine adapter AIR.
+pub trait VmAdapterInterface<T> {
+    /// The memory read data that should be exposed for downstream use
+    type Reads;
+    /// The memory write data that are expected to be provided by the integrator
+    type Writes;
+    /// The parts of the instruction that should be exposed to the integrator.
+    /// This will typically include `is_valid`, which indicates whether the trace row
+    /// is being used and `opcode` to indicate which opcode is being executed if the
+    /// VmChip supports multiple opcodes.
+    type ProcessedInstruction;
+}
+```
 
-The initial and final memory accesses are constrained different when the VM has continuations.
-See [Continuations](./continuations.md) for full details. In summary, because the initial and final memory states are
-committed to in a **trie**, the uniqueness of the addresses is constrained by the trie, so the arguments of the previous
-paragraph are not used.
+Here, `Reads` is some type that describes the memory reads required for the chip, and `Writes` is similar but about writes. The `VmAdapterChip` defines what the adapter needs to be able to do:
+
+```rust
```

Comment:
this kind of section with more specific code-level usage should go in `docs/crate` or in a README.md next to `integration_api.rs`

The "specs" markdown should be kept at a level above the code specifics, except in cases where you really need the code for precision. 

Since the integration API is only one of several ways to achieve what is laid out in our circuit architecture, I feel that it falls into the "crate docs" category and not part of the architecture spec.